backend/api/services.py

```diff
@@ -215,3 +221,172 @@ def get_reddit_posts(subreddit, symb, time="week") -> pd.DataFrame:
     df.replace("", np.nan, inplace=True)
     df.dropna(inplace=True)
     return df
+
+
+# TWEETS
+def get_tweets(query) -> pd.DataFrame:
+    """Returns a dataframe of tweets matching a given search query.
+
+    Args:
+        query (str): Search query eg. AAPL
+
+    Returns:
+        pd.DataFrame: DataFrame of tweets and other relevant information.
+    """
+    query = f"\\${query} lang:en"
+    tweets = twitter.search_recent_tweets(
+        query=query,
+        tweet_fields=["author_id", "created_at", "public_metrics", "entities"],
+        max_results=100,
+    )
+
+    tweets_df = pd.DataFrame(tweets.data)
+
+    # explode out public metrics
+    pub_metrics = tweets_df["public_metrics"].apply(pd.Series)
+
+    # extract hashtags
+    entities = pd.json_normalize(tweets_df["entities"])
+    hashtags = []
+    for i in entities["hashtags"]:
+        hashtags_str = ""
+        if isinstance(i, list):
+            for j in i:
+                hashtags_str += j["tag"] + " "
+        hashtags.append(hashtags_str)
+    hashtags = pd.Series(hashtags).rename("hashtags")
+
+    tweets_df = pd.concat([tweets_df, hashtags, pub_metrics], axis=1)
```

Comment:
Why explode out the `pub_metrics` just to add it back in again without any processing? Can you see if you can simplify the above code somehow?

Reply:
pub_metrics is a nested dict, that needs to be exploded out into columns. so this explodes it out, renames the new df as pub_metrics and concats it back to the original df

Reply:
got it